rust-symcrypt/src/rsa.rs

```diff
@@ -0,0 +1,479 @@
+/// My questions / Suggestions
+/// 1. We've introduced the idea of the RsaKeyPair state, which can hold a lot of data, and makes a lot of the calls to the underlying SymCrypt code redundant. most of the accessors no longer
+/// need to call SymCrypt since the same data is stored on the state.
+/// 2. The idea of new() then generate_key() / generate_set_public_only() / set_key_pair() is a bit convoluted IMO, we can squash these together to something like: 
+/// new_random(), and  new_set_public() and new_set_key_pair().
+/// 3. Many of the functions ( get, set etc ) take in a NumberFormat that has to match the other specified arguments, is this something we want to store on the state?
+/// 4. Another thought is to leave as is, and write a crate on top of this one, called something like symcrypt_rsa_parse, which will be a friendlier interface on top of the RSA code we have right now
+/// It would have convenience functions like rsa_from_pkcs8, rsa_from_der, rsa_from_asn1 etc. ( this could be part of the intern project as well.)
+/// 5. with how many steps / options there is for generating a key, we can maybe do a builder pattern, this would be similar to how it is on NCrypt / BCrypt where they have 
+/// RsaKey functions and the RsaKey.finalize() at the end.
+/// ex:
+///         builder = RsaKeyBuilder::new()
+///             .key_mode(RsaKeyMode::KeyPair)
+///             .key_usage(RsaKeyUsage::Sign)
+///             .generate_key()
+
+
+/// Impl for the RsaKeyPair
+impl RsaKeyPair {
+    /// `new()` allocates a the space needed for an `RsaKeyPair`.
+    /// To fill data, you must call one of [`RsaKeyPair::generate_key()`], [`RsaKeyPair::set_public_only()`] or [`RsaKeyPair::set_key_pair()`].
+    /// 
+    /// `n_bits_mod` represents the desired bit length of the Rsa Key.
+    /// 
+    /// `key_mode` will determine if this key will be allocated for a [`RsaKeyMode::PublicOnly`] or [`RsaKeyMode::KeyPair`] 
+    /// 
+    /// `key_usage` will indicate if this key will be used for [`RsaKeyUsage::Sign`] or [`RsaKeyUsage::Encrypt`]
+    pub fn new(n_bits_mod: u32, key_mode: RsaKeyMode, key_usage: RsaKeyUsage) -> Self {
+        let rsa_params = symcrypt_sys::SYMCRYPT_RSA_PARAMS {
+            version: 1 as symcrypt_sys::UINT32, // No other version aside from version 1 is specified
+            nBitsOfModulus: n_bits_mod as symcrypt_sys::UINT32,
+            nPrimes: key_mode.to_u32(), //  Value must be 0 or 2, if only a public key is being generated, you can pass 0 for n_primes
+            nPubExp: MAX_NUMBER_OF_PUBLIC_EXPONENTS, // Per SymCrypt documentation, only 1 is allowed
+        };
+        unsafe {
+            // SAFETY: FFI calls
+            // No flags specified for SymCryptRsakeyAllocate
+            let rsa_key = Box::pin(symcrypt_sys::SymCryptRsakeyAllocate(&rsa_params, 0));
```

Comment:
>Box::pin(symcrypt_sys::SymCryptRsakeyAllocate(&rsa_params, 0)); [](http://example.com/codeflow?start=26&length=63)

Is pinning necessary for allocated memory?

Reply:
The code works without the `Pin<Box<>>` but I don't know if rust can guarantee that the memory will not move after it's been allocated. Rust may or may not move things around under the covers:  https://doc.rust-lang.org/std/pin/

> Types that pin data to a location in memory.

It is sometimes useful to be able to rely upon a certain value not being able to move, in the sense that its address in memory cannot change. This is useful especially when there are one or more [pointers](https://doc.rust-lang.org/std/primitive.pointer.html) pointing at that value. The ability to rely on this guarantee that the value a [pointer](https://doc.rust-lang.org/std/primitive.pointer.html) is pointing at (its pointee) will

Not be moved out of its memory location
More generally, remain valid at that same memory location

> When we say a value is moved, we mean that the compiler copies, byte-for-byte, the value from one location to another. In a purely mechanical sense, this is identical to [Copy](https://doc.rust-lang.org/std/marker/trait.Copy.html)ing a value from one place in memory to another. In Rust, “move” carries with it the semantics of ownership transfer from one variable to another, which is the key difference between a [Copy](https://doc.rust-lang.org/std/marker/trait.Copy.html) and a move. For the purposes of this module’s documentation, however, when we write move in italics, we mean specifically that the value has moved in the mechanical sense of being located at a new place in memory.


I think that if we removed the `Pin<Box<>>` it would probably work, but id rather be safe than sorry in this scenario